homeassistant/components/zwave_mqtt/entity.py

```diff
@@ -0,0 +1,272 @@
+            # Skip if it's already been added.
+            if self._values[name] is not None:
+                continue
+            # Skip if the value doesn't match the schema.
+            if not check_value_schema(value, self._schema[const.DISC_VALUES][name]):
+                continue
+
+            # Add value to mapping.
+            self._values[name] = value
+
+            # If the entity has already been created, notify it of the new value.
+            if self._entity_created:
+                async_dispatcher_send(self._hass, f"{self.values_id}_value_added")
+
+            # Check if entity has all required values and create the entity if needed.
+            self._check_entity_ready()
+
+    @callback
+    def _check_entity_ready(self):
+        """Check if all required values are discovered and create entity."""
+        # Abort if the entity has already been created
+        if self._entity_created:
+            return
+
+        # Go through values defined in the schema and abort if a required value is missing.
+        for name, disc_settings in self._schema[const.DISC_VALUES].items():
+            if self._values[name] is None and not disc_settings.get(
+                const.DISC_OPTIONAL
+            ):
+                return
+
+        # We have all the required values, so create the entity.
+        component = self._schema[const.DISC_COMPONENT]
+
+        _LOGGER.debug(
+            "Adding Node_id=%s Generic_command_class=%s, "
+            "Specific_command_class=%s, "
+            "Command_class=%s, Index=%s, Value type=%s, "
+            "Genre=%s as %s",
+            self._node.node_id,
+            self._node.node_generic,
+            self._node.node_specific,
+            self.primary.command_class,
+            self.primary.index,
+            self.primary.type,
+            self.primary.genre,
+            component,
+        )
+        self._entity_created = True
+
+        if component in PLATFORMS:
+            async_dispatcher_send(self._hass, f"zwave_new_{component}", self)
+
+    @property
+    def values_id(self):
+        """Identification for this values collection."""
+        return f"{self.primary.node.id}-{self.primary.value_id_key}"
+
+
+class ZWaveDeviceEntity(Entity):
+    """Generic Entity Class for a Z-Wave Device."""
+
+    def __init__(self, values):
+        """Initialize a generic Z-Wave device entity."""
+        self.values = values
+        self.options = values.options
+
+    @callback
+    def on_value_update(self):
+        """Call when a value is added/updated in the entity EntityValues Collection.
+
+        To be overridden by platforms needing this event.
+        """
+
+    async def async_added_to_hass(self):
+        """Call when entity is added."""
+        # add dispatcher and OZW listeners callbacks,
+        self.options.listen(EVENT_VALUE_CHANGED, self._value_changed)
+        self.options.listen(EVENT_INSTANCE_STATUS_CHANGED, self._instance_updated)
```

Comment:
We should also unlisten these on entity removal.

Reply:
That's done already. Dirty but done.